join_group_request.go

```diff
@@ -0,0 +1,111 @@
+package sarama
+
+type JoinGroupRequest struct {
+	GroupId        string
+	SessionTimeout int32
+	MemberId       string
+	ProtocolType   string
+	GroupProtocols []GroupProtocol
+}
+
+func (r *JoinGroupRequest) encode(pe packetEncoder) error {
+	if err := pe.putString(r.GroupId); err != nil {
+		return err
+	}
+	pe.putInt32(r.SessionTimeout)
+	if err := pe.putString(r.MemberId); err != nil {
+		return err
+	}
+	if err := pe.putString(r.ProtocolType); err != nil {
+		return err
+	}
+
+	if err := pe.putArrayLength(len(r.GroupProtocols)); err != nil {
+		return err
+	}
+	for _, groupProtocol := range r.GroupProtocols {
+		if err := groupProtocol.encodeGroupProtocol(pe); err != nil {
+			return err
+		}
+	}
+
+	return nil
+}
+
+func (r *JoinGroupRequest) decode(pd packetDecoder) (err error) {
+	r.GroupId, err = pd.getString()
+	if err != nil {
+		return
+	}
+
+	r.SessionTimeout, err = pd.getInt32()
+	if err != nil {
+		return
+	}
+
+	r.MemberId, err = pd.getString()
+	if err != nil {
+		return
+	}
+
+	r.ProtocolType, err = pd.getString()
+	if err != nil {
+		return
+	}
+
+	switch r.ProtocolType {
+	case "consumer":
+		n, err := pd.getArrayLength()
+		if err != nil {
+			return err
+		}
+
+		r.GroupProtocols = make([]GroupProtocol, n)
+		for i := 0; i < n; i++ {
+			r.GroupProtocols[i] = new(ConsumerGroupProtocol)
+			if err := r.GroupProtocols[i].decodeGroupProtocol(pd); err != nil {
+				return nil
+			}
+		}
+
+	default:
+		return ErrUnknownGroupProtocol
+	}
+
+	return nil
+}
+
+func (r *JoinGroupRequest) key() int16 {
+	return 11
+}
+
+func (r *JoinGroupRequest) version() int16 {
+	return 0
+}
+
+type GroupProtocol interface {
+	encodeGroupProtocol(packetEncoder) error
+	decodeGroupProtocol(packetDecoder) error
+}
```

Comment:
This part is a bit tricky; the protocol defines this to be extensible based on the `ProtocolType`. It continues to only define the specifics for the "consumer" protocol.